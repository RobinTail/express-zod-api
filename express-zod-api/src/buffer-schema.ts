import { z } from "zod/v4";

export const bufferSchema = z.custom<Buffer>(
  (subject) => Buffer.isBuffer(subject),
  { error: "Expected Buffer" },
);
