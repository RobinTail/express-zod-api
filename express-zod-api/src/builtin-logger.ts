import ansis from "ansis";
import { inspect } from "node:util";
import { performance } from "node:perf_hooks";
import { FlatObject, isProduction } from "./common-helpers";
import {
  AbstractLogger,
  formatDuration,
  isHidden,
  Severity,
  styles,
} from "./logger-helpers";

interface Context extends FlatObject {
  requestId?: string;
}

export interface BuiltinLoggerConfig {
  /**
   * @desc The minimal severity to log or "silent" to disable logging
   * @example "debug" also enables pretty output for inspected entities
   * */
  level: "silent" | "warn" | "info" | "debug";
  /** @desc Enables colors on printed severity and inspected entities */
  color: boolean;
  /**
   * @desc Printing logs does not delay execution
   * @see https://expressjs.com/en/advanced/best-practice-performance.html#do-logging-correctly
   * */
  async: boolean;
  /**
   * @desc Control how deeply entities should be inspected
   * @example null
   * @example Infinity
   * */
  depth: number | null;
  /**
   * @desc Context: the metadata applicable for each logged entry, used by .child() method
   * @see childLoggerProvider
   * */
  ctx: Context;
}

interface ProfilerOptions {
  message: string;
  /** @default "debug" */
  severity?: Severity | ((ms: number) => Severity);
  /** @default formatDuration - adaptive units and limited fraction */
  formatter?: (ms: number) => string | number;
}

/** @desc Built-in console logger with optional colorful inspections */
export class BuiltinLogger implements AbstractLogger {
  protected readonly config: BuiltinLoggerConfig;
  protected readonly stack: string[] = [];
  protected postponed?: NodeJS.Immediate;

  /** @example new BuiltinLogger({ level: "debug", color: true, depth: 4 }) */
  public constructor({
    color = ansis.isSupported(),
    level = isProduction() ? "warn" : "debug",
    async = isProduction(),
    depth = 2,
    ctx = {},
  }: Partial<BuiltinLoggerConfig> = {}) {
    this.config = { color, level, async, depth, ctx };
  }

  protected format(subject: unknown) {
    const { depth, color: colors, level } = this.config;
    return inspect(subject, {
      depth,
      colors,
      breakLength: level === "debug" ? 80 : Infinity,
      compact: level === "debug" ? 3 : true,
    });
  }

  protected print(method: Severity, message: string, meta?: unknown) {
    const {
      level,
      ctx: { requestId, ...ctx },
      color: hasColor,
      async: isAsync,
    } = this.config;
    if (level === "silent" || isHidden(method, level)) return;
    const output: string[] = [new Date().toISOString()];
    if (requestId) output.push(hasColor ? styles.ctx(requestId) : requestId);
    output.push(
      hasColor ? `${styles[method](method)}:` : `${method}:`,
      message,
    );
    if (meta !== undefined) output.push(this.format(meta));
    if (Object.keys(ctx).length > 0) output.push(this.format(ctx));
    (isAsync ? this.postpone.bind(this) : console.log)(output.join(" "));
  }

  protected postpone(line: string) {
    this.stack.push(line);
    clearImmediate(this.postponed);
    this.postponed = setImmediate(this.purge.bind(this)).unref(); // do not block process.exit()
  }

  /** @internal */
  public purge() {
    while (this.stack.length) console.log(this.stack.shift());
  }

  /** @internal */
  public dispose(cb: () => void) {
    this.config.level = "silent"; // no more logs
    clearImmediate(this.postponed);
    this.postponed = setImmediate(() => {
      this.purge();
      cb();
    });
  }

  public debug(message: string, meta?: unknown) {
    this.print("debug", message, meta);
  }

  public info(message: string, meta?: unknown) {
    this.print("info", message, meta);
  }

  public warn(message: string, meta?: unknown) {
    this.print("warn", message, meta);
  }

  public error(message: string, meta?: unknown) {
    this.print("error", message, meta);
  }

  public child(ctx: Context) {
    return new BuiltinLogger({ ...this.config, ctx });
  }

  /**
   * @desc The argument used for instance created by .child() method
   * @see ChildLoggerProvider
   * */
  public get ctx() {
    return this.config.ctx;
  }

  /** @desc Measures the duration until you invoke the returned callback */
  public profile(message: string): () => void;
  public profile(options: ProfilerOptions): () => void;
  public profile(subject: string | ProfilerOptions) {
    const start = performance.now();
    return () => {
      const duration = performance.now() - start;
      const {
        message,
        severity = "debug",
        formatter = formatDuration,
      } = typeof subject === "object" ? subject : { message: subject };
      this.print(
        typeof severity === "function" ? severity(duration) : severity,
        message,
        formatter(duration),
      );
    };
  }
}
