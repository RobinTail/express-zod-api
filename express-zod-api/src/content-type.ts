export const contentTypes = {
  json: "application/json",
  upload: "multipart/form-data",
  raw: "application/octet-stream",
  sse: "text/event-stream",
};

export type ContentType = keyof typeof contentTypes;
export type RequestType = Extract<ContentType, "json" | "upload" | "raw">;
