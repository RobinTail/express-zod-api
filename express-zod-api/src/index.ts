import "@express-zod-api/zod-plugin"; // side effects here
export { createConfig } from "./config-type.ts";
export {
  EndpointsFactory,
  defaultEndpointsFactory,
  arrayEndpointsFactory,
} from "./endpoints-factory.ts";
export { getMessageFromError } from "./common-helpers.ts";
export { ensureHttpError } from "./result-helpers.ts";
export { BuiltinLogger } from "./builtin-logger.ts";
export { Middleware } from "./middleware.ts";
export {
  ResultHandler,
  defaultResultHandler,
  arrayResultHandler,
} from "./result-handler.ts";
export { DependsOnMethod } from "./depends-on-method.ts";
export { ServeStatic } from "./serve-static.ts";
export { createServer, attachRouting } from "./server.ts";
export { Documentation } from "./documentation.ts";
export {
  DocumentationError,
  RoutingError,
  OutputValidationError,
  InputValidationError,
  MissingPeerError,
} from "./errors.ts";
export { testEndpoint, testMiddleware } from "./testing.ts";
export { Integration } from "./integration.ts";
export { EventStreamFactory } from "./sse.ts";

export { ez } from "./proprietary-schemas.ts";

// Convenience types
export type { Depicter } from "./documentation-helpers.ts";
export type { Producer } from "./zts-helpers.ts";

// Interfaces exposed for augmentation
export type { LoggerOverrides } from "./logger-helpers.ts";
export type { TagOverrides } from "./common-helpers.ts";

// Issues 952, 1182, 1269: Insufficient exports for consumer's declaration
import type {} from "qs"; // fixes TS2742 for attachRouting
export type { Routing } from "./routing.ts";
export type { FlatObject } from "./common-helpers.ts";
export type { Method } from "./method.ts";
export type { IOSchema } from "./io-schema.ts";
export type { CommonConfig, AppConfig, ServerConfig } from "./config-type.ts";
export type { ApiResponse } from "./api-response.ts";
export type {
  BasicSecurity,
  BearerSecurity,
  CookieSecurity,
  HeaderSecurity,
  InputSecurity,
  OAuth2Security,
  OpenIdSecurity,
} from "./security.ts";
