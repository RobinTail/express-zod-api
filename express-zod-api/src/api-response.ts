import { z } from "zod";
import { isSchema } from "./common-helpers";
import { contentTypes } from "./content-type";

export const defaultStatusCodes = {
  positive: 200,
  negative: 400,
} satisfies Record<string, number>;

export type ResponseVariant = keyof typeof defaultStatusCodes;
export const responseVariants = Object.keys(
  defaultStatusCodes,
) as ResponseVariant[];

/** @public this is the user facing configuration */
export class ApiResponse<S extends z.ZodType> {
  readonly #schema: S;
  readonly #statusCodes?: [number, ...number[]];
  readonly #mimeTypes?: [string, ...string[]] | null;

  public constructor(
    subject:
      | S // shorthand for { schema: S }, default status codes and MIME types applied
      | {
          schema: S;
          /** @default 200 for a positive and 400 for a negative response */
          statusCode?: number | [number, ...number[]];
          /**
           * @example null is for no content, such as 204 and 302
           * @default "application/json"
           * */
          mimeType?: string | [string, ...string[]] | null;
        },
  ) {
    const { schema, statusCode, mimeType } = isSchema(subject)
      ? { schema: subject }
      : subject;
    this.#schema = schema;
    this.#statusCodes =
      typeof statusCode === "number" ? [statusCode] : statusCode;
    this.#mimeTypes = typeof mimeType === "string" ? [mimeType] : mimeType;
  }

  public normalize(fbStatus: number) {
    return {
      schema: this.#schema,
      statusCodes: this.#statusCodes ?? [fbStatus],
      mimeTypes:
        this.#mimeTypes === undefined ? [contentTypes.json] : this.#mimeTypes,
    };
  }
}

export type NormalizedResponse = ReturnType<
  ApiResponse<z.ZodType>["normalize"]
>;
