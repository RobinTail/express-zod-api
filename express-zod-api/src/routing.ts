import { IRouter, RequestHandler } from "express";
import createHttpError from "http-errors";
import { isProduction } from "./common-helpers";
import { CommonConfig } from "./config-type";
import { ContentType } from "./content-type";
import { Diagnostics } from "./diagnostics";
import { AuxMethod, Method } from "./method";
import { Routable } from "./routable";
import { OnEndpoint, walkRouting } from "./routing-walker";
import { GetLogger } from "./server-helpers";

export interface Routing {
  [SEGMENT: string]: Routing | Routable;
}

export type Parsers = Partial<Record<ContentType, RequestHandler[]>>;

/** @link https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/405 */
export const createWrongMethodHandler =
  (allowedMethods: Array<Method | AuxMethod>): RequestHandler =>
  ({ method }, res, next) => {
    const Allow = allowedMethods.join(", ").toUpperCase();
    res.set({ Allow }); // in case of a custom errorHandler configured that does not care about headers in error
    const error = createHttpError(405, `${method} is not allowed`, {
      headers: { Allow },
    });
    next(error);
  };

export const initRouting = ({
  app,
  getLogger,
  config,
  routing,
  parsers,
}: {
  app: IRouter;
  getLogger: GetLogger;
  config: CommonConfig;
  routing: Routing;
  parsers?: Parsers;
}) => {
  let doc: Diagnostics | undefined = new Diagnostics(getLogger()); // disposable instance
  const familiar = new Map<string, Array<Method | AuxMethod>>();
  const onEndpoint: OnEndpoint = (endpoint, path, method, siblingMethods) => {
    if (!isProduction()) {
      doc?.checkJsonCompat(endpoint, { path, method });
      doc?.checkPathParams(path, endpoint, { method });
    }
    const matchingParsers = parsers?.[endpoint.getRequestType()] || [];
    const handler: RequestHandler = async (request, response) => {
      const logger = getLogger(request);
      if (config.cors) {
        const accessMethods: Array<Method | AuxMethod> = [
          method,
          ...(siblingMethods || []),
          "options",
        ];
        const methodsLine = accessMethods.join(", ").toUpperCase();
        const defaultHeaders: Record<string, string> = {
          "Access-Control-Allow-Origin": "*",
          "Access-Control-Allow-Methods": methodsLine,
          "Access-Control-Allow-Headers": "content-type",
        };
        const headers =
          typeof config.cors === "function"
            ? await config.cors({ request, endpoint, logger, defaultHeaders })
            : defaultHeaders;
        for (const key in headers) response.set(key, headers[key]);
      }
      return endpoint.execute({ request, response, logger, config });
    };
    if (!familiar.has(path)) {
      familiar.set(path, []);
      if (config.cors) {
        app.options(path, ...matchingParsers, handler);
        familiar.get(path)?.push("options");
      }
    }
    familiar.get(path)?.push(method);
    app[method](path, ...matchingParsers, handler);
  };
  walkRouting({ routing, onEndpoint, onStatic: app.use.bind(app) });
  doc = undefined; // hint for garbage collector
  if (config.wrongMethodBehavior !== 405) return;
  for (const [path, allowedMethods] of familiar.entries())
    app.all(path, createWrongMethodHandler(allowedMethods));
};
