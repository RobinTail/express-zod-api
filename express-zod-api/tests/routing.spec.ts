import {
  appMock,
  expressMock,
  staticHandler,
  staticMock,
} from "./express-mock";
import { z } from "zod";
import {
  EndpointsFactory,
  Routing,
  ServeStatic,
  defaultResultHandler,
  ez,
} from "../src";
import {
  makeLoggerMock,
  makeRequestMock,
  makeResponseMock,
} from "../src/testing";
import { createWrongMethodHandler, initRouting } from "../src/routing";
import type { IRouter, RequestHandler } from "express";
import createHttpError from "http-errors";

describe("Routing", () => {
  describe("initRouting()", () => {
    beforeAll(() => {
      expressMock();
    });

    beforeEach(() => {
      vi.clearAllMocks(); // resets call counters on mocked methods
    });

    test.each([404, 405] as const)(
      "Should set right methods %#",
      (wrongMethodBehavior) => {
        const handlerMock = vi.fn();
        const configMock = {
          cors: true,
          startupLogo: false,
          wrongMethodBehavior,
        };
        const factory = new EndpointsFactory(defaultResultHandler);
        const getEndpoint = factory.build({
          output: z.object({}),
          handler: handlerMock,
        });
        const postEndpoint = factory.build({
          method: "post",
          output: z.object({}),
          handler: handlerMock,
        });
        const getAndPostEndpoint = factory.build({
          method: ["get", "post"],
          output: z.object({}),
          handler: handlerMock,
        });
        const routing: Routing = {
          v1: {
            user: {
              get: getEndpoint,
              set: postEndpoint,
              universal: getAndPostEndpoint,
            },
          },
        };
        const logger = makeLoggerMock();
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: configMock,
          routing,
        });
        expect(appMock.get).toHaveBeenCalledTimes(2);
        expect(appMock.post).toHaveBeenCalledTimes(2);
        expect(appMock.put).toHaveBeenCalledTimes(0);
        expect(appMock.delete).toHaveBeenCalledTimes(0);
        expect(appMock.patch).toHaveBeenCalledTimes(0);
        expect(appMock.options).toHaveBeenCalledTimes(3);
        expect(appMock.get.mock.calls[0][0]).toBe("/v1/user/get");
        expect(appMock.get.mock.calls[1][0]).toBe("/v1/user/universal");
        expect(appMock.post.mock.calls[0][0]).toBe("/v1/user/set");
        expect(appMock.post.mock.calls[1][0]).toBe("/v1/user/universal");
        expect(appMock.options.mock.calls[0][0]).toBe("/v1/user/get");
        expect(appMock.options.mock.calls[1][0]).toBe("/v1/user/set");
        expect(appMock.options.mock.calls[2][0]).toBe("/v1/user/universal");
        if (wrongMethodBehavior !== 405) return;
        expect(appMock.all).toHaveBeenCalledTimes(3);
        expect(appMock.all.mock.calls[0][0]).toBe("/v1/user/get");
        expect(appMock.all.mock.calls[1][0]).toBe("/v1/user/set");
        expect(appMock.all.mock.calls[2][0]).toBe("/v1/user/universal");
      },
    );

    test("Should accept serveStatic", () => {
      const routing: Routing = {
        public: new ServeStatic(__dirname, { dotfiles: "deny" }),
      };
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: true },
        routing,
      });
      expect(staticMock).toHaveBeenCalledWith(__dirname, { dotfiles: "deny" });
      expect(appMock.use).toHaveBeenCalledTimes(1);
      expect(appMock.use).toHaveBeenCalledWith("/public", staticHandler);
    });

    test("Should handle method depending assignments", () => {
      const handlerMock = vi.fn();
      const factory = new EndpointsFactory(defaultResultHandler);
      const getEndpoint = factory.build({
        output: z.object({}),
        handler: handlerMock,
      });
      const postEndpoint = factory.build({
        output: z.object({}),
        handler: handlerMock,
      });
      const putAndPatchEndpoint = factory.build({
        output: z.object({}),
        handler: handlerMock,
      });
      const routing: Routing = {
        v1: {
          user: {
            "get /": getEndpoint,
            "post /": postEndpoint,
            "put /": putAndPatchEndpoint,
            "patch /": putAndPatchEndpoint,
          },
        },
      };
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: true },
        routing,
      });
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.post).toHaveBeenCalledTimes(1);
      expect(appMock.put).toHaveBeenCalledTimes(1);
      expect(appMock.patch).toHaveBeenCalledTimes(1);
      expect(appMock.delete).toHaveBeenCalledTimes(0);
      expect(appMock.options).toHaveBeenCalledTimes(1);
      expect(appMock.get.mock.calls[0][0]).toBe("/v1/user");
      expect(appMock.post.mock.calls[0][0]).toBe("/v1/user");
      expect(appMock.put.mock.calls[0][0]).toBe("/v1/user");
      expect(appMock.patch.mock.calls[0][0]).toBe("/v1/user");
      expect(appMock.options.mock.calls[0][0]).toBe("/v1/user");
    });

    test("Should check if endpoint supports the method it's assigned to", () => {
      const factory = new EndpointsFactory(defaultResultHandler);
      const putAndPatchEndpoint = factory.build({
        method: ["put", "patch"],
        output: z.object({}),
        handler: vi.fn(),
      });
      const routing: Routing = {
        v1: {
          user: {
            "put /": putAndPatchEndpoint,
            "patch /": putAndPatchEndpoint,
            "post /": putAndPatchEndpoint, // intentional
          },
        },
      };
      const logger = makeLoggerMock();
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: true },
          routing,
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Issue 705: should set all assigned methods to CORS response header", async () => {
      const handler = vi.fn(async () => ({}));
      const configMock = {
        cors: (params: { defaultHeaders: Record<string, string> }) => ({
          ...params.defaultHeaders,
          "X-Custom-Header": "Testing",
        }),
      };
      const factory = new EndpointsFactory(defaultResultHandler);
      const input = z.object({});
      const output = z.object({});
      const getEndpoint = factory.build({
        input,
        output,
        handler,
      });
      const postEndpoint = factory.build({
        input,
        output,
        handler,
      });
      const putAndPatchEndpoint = factory.build({
        input,
        output,
        handler,
      });
      const routing: Routing = {
        hello: {
          "get /": getEndpoint,
          "post /": postEndpoint,
          "put /": putAndPatchEndpoint,
          "patch /": putAndPatchEndpoint,
        },
      };
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: configMock,
        routing,
      });
      expect(appMock.options).toHaveBeenCalledTimes(1);
      expect(appMock.options.mock.calls[0][0]).toBe("/hello");
      const fn = appMock.options.mock.calls[0][1];
      expect(typeof fn).toBe("function"); // async RequestHandler, proprietary CORS middleware
      const requestMock = makeRequestMock({ method: "PUT" });
      const responseMock = makeResponseMock();
      await fn(requestMock, responseMock, vi.fn());
      expect(responseMock._getStatusCode()).toBe(200);
      expect(responseMock._getHeaders()).toEqual({
        "access-control-allow-origin": "*",
        "access-control-allow-methods": "GET, PATCH, POST, PUT, HEAD, OPTIONS",
        "access-control-allow-headers": "content-type",
        "x-custom-header": "Testing",
      });
    });

    test("Should accept parameters", () => {
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        output: z.object({}),
        handler: vi.fn(),
      });
      const routing: Routing = {
        v1: {
          user: {
            ":id": endpointMock,
          },
        },
      };
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing,
      });
      expect(appMock.get).toHaveBeenCalledTimes(1);
      expect(appMock.get.mock.calls[0][0]).toBe("/v1/user/:id");
    });

    test("Should handle empty paths and trim spaces", () => {
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        output: z.object({}),
        handler: vi.fn(),
      });
      const routing: Routing = {
        v1: {
          user: {
            ":id": endpointMock.nest({
              " download ": endpointMock,
            }),
          },
        },
      };
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing,
      });
      expect(appMock.get).toHaveBeenCalledTimes(2);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/user/:id",
        expect.any(Function),
      );
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/user/:id/download",
        expect.any(Function),
      );
    });

    test("Should handle slashes in routing keys", () => {
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        output: z.object({}),
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing: {
          v1: { "///user/retrieve///": endpointMock },
          "v1/user/delete": endpointMock,
        },
      });
      expect(appMock.get).toHaveBeenCalledTimes(2);
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/user/retrieve",
        expect.any(Function),
      );
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/user/delete",
        expect.any(Function),
      );
    });

    test("Should handle explicitly specified method", async () => {
      const endpoint = new EndpointsFactory(defaultResultHandler).buildVoid({
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: true },
        routing: {
          v1: {
            "get ///user/retrieve///": endpoint,
            user: {
              "post retrieve": endpoint,
            },
          },
        },
      });
      expect(appMock.get).toHaveBeenCalledOnce();
      expect(appMock.get).toHaveBeenCalledWith(
        "/v1/user/retrieve",
        expect.any(Function), // cors
        expect.any(Function), // endpoint
      );
      expect(appMock.post).toHaveBeenCalledOnce();
      expect(appMock.post).toHaveBeenCalledWith(
        "/v1/user/retrieve",
        expect.any(Function), // cors
        expect.any(Function), // endpoint
      );
      expect(appMock.options).toHaveBeenCalledTimes(1);
      expect(appMock.options.mock.calls[0]).toEqual([
        "/v1/user/retrieve",
        expect.any(Function), // cors
        expect.any(Function), // endpoint
      ]);
      const fn = appMock.options.mock.calls[0][1]; // similar to issue 705
      const requestMock = makeRequestMock({ method: "POST" });
      const responseMock = makeResponseMock();
      await fn(requestMock, responseMock, vi.fn());
      expect(responseMock._getStatusCode()).toBe(200);
      expect(responseMock._getHeaders()).toEqual({
        "access-control-allow-origin": "*",
        "access-control-allow-methods": "GET, POST, HEAD, OPTIONS",
        "access-control-allow-headers": "content-type",
      });
    });

    test("Should check if endpoint supports an explicitly specified method", () => {
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        method: "post",
        output: z.object({}),
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: false },
          routing: {
            v1: {
              "get ///user/retrieve///": endpointMock,
            },
          },
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Should prohibit nested routing within a route having explicit method", () => {
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        output: z.object({}),
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: false },
          routing: {
            v1: {
              "get ///user/retrieve///": { another: endpointMock },
            },
          },
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Should prohibit nesting for a route having explicit method", () => {
      const logger = makeLoggerMock();
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: false },
          routing: {
            v1: {
              "get /user/retrieve": {},
            },
          },
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Should prohibit ServeStatic for a route having explicit method", () => {
      const logger = makeLoggerMock();
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: false },
          routing: {
            v1: {
              "get /user/retrieve": new ServeStatic("."),
            },
          },
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Should prohibit duplicated routes", () => {
      const logger = makeLoggerMock();
      const endpointMock = new EndpointsFactory(defaultResultHandler).build({
        output: z.object({}),
        handler: vi.fn(),
      });
      const routing: Routing = {
        v1: { test: endpointMock },
        "/v1/test": endpointMock,
      };
      expect(() =>
        initRouting({
          app: appMock as unknown as IRouter,
          getLogger: () => logger,
          config: { cors: false },
          routing,
        }),
      ).toThrowErrorMatchingSnapshot();
    });

    test("Should execute endpoints with right arguments", async () => {
      const handlerMock = vi
        .fn()
        .mockImplementationOnce(() => ({ result: true }));
      const setEndpoint = new EndpointsFactory(defaultResultHandler).build({
        method: "post",
        input: z.object({ test: z.number() }),
        output: z.object({ result: z.boolean() }),
        handler: handlerMock,
      });
      const routing: Routing = {
        v1: {
          user: {
            set: setEndpoint,
          },
        },
      };
      const getLoggerMock = vi.fn(() => makeLoggerMock());
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: getLoggerMock,
        config: { cors: true },
        routing,
      });
      expect(appMock.post).toHaveBeenCalledTimes(1);
      const routeHandler = appMock.post.mock.calls[0][2] as RequestHandler; // 1 is CORS
      const requestMock = makeRequestMock({
        method: "POST",
        body: { test: 123 },
      });
      const responseMock = makeResponseMock();
      const nextMock = vi.fn();
      await routeHandler(requestMock, responseMock, nextMock);
      expect(getLoggerMock).toHaveBeenCalledWith(requestMock);
      expect(nextMock).toHaveBeenCalledTimes(0);
      expect(handlerMock).toHaveBeenCalledTimes(1);
      expect(
        getLoggerMock.mock.results.pop()!.value._getLogs().error,
      ).toHaveLength(0);
      expect(handlerMock).toHaveBeenCalledWith({
        input: { test: 123 },
        options: {},
        logger: getLoggerMock.mock.results.pop()!.value,
      });
      expect(responseMock._getStatusCode()).toBe(200);
      expect(responseMock._getJSONData()).toEqual({
        status: "success",
        data: { result: true },
      });
    });

    const circular: z.ZodType = z.lazy(() => z.tuple([circular, z.nan()]));
    test.each([
      [z.bigint(), z.set(z.string())],
      [z.nan(), z.map(z.string(), z.boolean())],
      [
        z.date().transform(String).pipe(z.string()),
        z.symbol().catch(Symbol("test")),
      ],
      [ez.dateOut(), ez.dateIn()],
      [z.lazy(() => z.void()), ez.raw()],
      [z.promise(z.any()), ez.upload()],
      [z.never(), z.tuple([ez.buffer()]).rest(z.nan())],
      [ez.buffer().pipe(z.any()), circular],
    ])("should warn about JSON incompatible schemas %#", (input, output) => {
      const endpoint = new EndpointsFactory(defaultResultHandler).build({
        input: z.object({ input }),
        output: z.object({ output }),
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing: { path: endpoint },
      });
      expect(logger._getLogs().warn).toEqual([
        [
          "The final input schema of the endpoint contains an unsupported JSON payload type.",
          { method: "get", path: "/path", reason: expect.any(z.ZodType) },
        ],
        [
          "The final positive response schema of the endpoint contains an unsupported JSON payload type.",
          { method: "get", path: "/path", reason: expect.any(z.ZodType) },
        ],
      ]);
    });

    test.each([
      [z.string().array(), z.string()],
      [
        z.lazy(() => z.number()),
        z.record(z.number(), z.string()).pipe(z.array(z.string())),
      ],
    ])("should warn about non-object based schemas I/O %#", (input, output) => {
      const endpoint = new EndpointsFactory(defaultResultHandler).build({
        input: input as unknown as z.ZodObject,
        output: output as unknown as z.ZodObject,
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing: { path: endpoint },
      });
      expect(logger._getLogs().warn).toEqual([
        [
          "Endpoint input schema is not object-based",
          { method: "get", path: "/path" },
        ],
        [
          "Endpoint output schema is not object-based",
          { method: "get", path: "/path" },
        ],
      ]);
    });

    test.each([
      z.object({ id: z.string() }),
      z.record(z.literal("id"), z.string()),
    ])("should warn about unused path params %#", (input) => {
      const endpoint = new EndpointsFactory(defaultResultHandler).build({
        input,
        output: z.object({}),
        handler: vi.fn(),
      });
      const logger = makeLoggerMock();
      initRouting({
        app: appMock as unknown as IRouter,
        getLogger: () => logger,
        config: { cors: false },
        routing: { v1: { ":idx": endpoint } },
      });
      expect(logger._getLogs().warn).toContainEqual([
        "The input schema of the endpoint is most likely missing the parameter of the path it's assigned to.",
        { method: "get", param: "idx", path: "/v1/:idx" },
      ]);
    });
  });

  describe("createWrongMethodHandler", () => {
    test("should call forward 405 error with a header having list of allowed methods", () => {
      const handler = createWrongMethodHandler(["post", "options"]);
      const nextMock = vi.fn();
      const resMock = makeResponseMock();
      handler(makeRequestMock(), resMock, nextMock);
      expect(resMock._getHeaders()).toHaveProperty("allow", "POST, OPTIONS");
      expect(nextMock).toHaveBeenCalledWith(
        createHttpError(405, "GET is not allowed", {
          headers: { Allow: "POST, OPTIONS" },
        }),
      );
    });
  });
});
