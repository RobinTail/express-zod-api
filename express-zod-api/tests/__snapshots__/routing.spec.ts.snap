// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Routing > initRouting() > Should check if endpoint supports an explicitly specified method 1`] = `
RoutingError({
  "cause": {
    "method": "get",
    "path": "/v1/user/retrieve",
  },
  "message": "Method get is not supported by the assigned Endpoint.",
})
`;

exports[`Routing > initRouting() > Should check if endpoint supports the method it's assigned to within DependsOnMethod 1`] = `
RoutingError({
  "cause": {
    "method": "post",
    "path": "/v1/user",
  },
  "message": "Method post is not supported by the assigned Endpoint.",
})
`;

exports[`Routing > initRouting() > Should prohibit DependsOnMethod for a route having explicit method 1`] = `
RoutingError({
  "cause": {
    "method": "get",
    "path": "/v1/user/retrieve",
  },
  "message": "Explicitly specified method can not be combined with DependsOnMethod.",
})
`;

exports[`Routing > initRouting() > Should prohibit ServeStatic for a route having explicit method 1`] = `
RoutingError({
  "cause": {
    "method": "get",
    "path": "/v1/user/retrieve",
  },
  "message": "Explicitly specified method can not be combined with ServeStatic.",
})
`;

exports[`Routing > initRouting() > Should prohibit duplicated routes 1`] = `
RoutingError({
  "cause": {
    "method": "get",
    "path": "/v1/test",
  },
  "message": "Route has a duplicate",
})
`;

exports[`Routing > initRouting() > Should prohibit nested routing within a route having explicit method 1`] = `
RoutingError({
  "cause": {
    "method": "get",
    "path": "/v1/user/retrieve",
  },
  "message": "Explicitly specified method can not be combined with nested routing.",
})
`;
